src/ice_usb_uart.c

```diff
@@ -0,0 +1,245 @@
+//
+// Created by MrJake222 on 03/11/24.
+//
+
+#include "ice_usb_uart.h"
+
+#include "rb.h"
+#include "hardware/uart.h"
+#include "hardware/irq.h"
+
+struct uart_wrap {
+    int itf;            // CDC interface number
+    uart_inst_t* inst;  // pico sdk instance reference
+
+    int irq_num;
+    void (*irq_handler)();
+
+    struct rb rx_buf;
+    struct rb tx_buf;
+
+    absolute_time_t dbg_time;
+};
+
+
+// ------------------------------- SDK extensions -------------------------------
+enum uart_watermark {
+    // uart fifo is 32 bytes deep
+    UART_WATERMARK_4CHR = 0,
+    UART_WATERMARK_8CHR = 1,
+    UART_WATERMARK_16CHR = 2,
+    UART_WATERMARK_24CHR = 3,
+    UART_WATERMARK_28CHR = 4
+};
+
+void uart_set_irq_watermarks(struct uart_wrap* uart, enum uart_watermark value) {
+    hw_write_masked(&uart_get_hw(uart->inst)->ifls, value << UART_UARTIFLS_RXIFLSEL_LSB,
+                    UART_UARTIFLS_RXIFLSEL_BITS);
+
+    hw_write_masked(&uart_get_hw(uart->inst)->ifls, value << UART_UARTIFLS_TXIFLSEL_LSB,
+                    UART_UARTIFLS_TXIFLSEL_BITS);
+}
+
+// ------------------------------- Rx functions -------------------------------
+// uart -> rb (irq)
+//   rb -> cdc
+
+static void ice_usb_uart_rx_irq(struct uart_wrap* uart) {
+    while (uart_is_readable(uart->inst)) {
+        uint8_t byte = uart_getc(uart->inst);
+        *rb_get_write_ptr(&uart->rx_buf) = byte;
+        rb_write_ack(&uart->rx_buf, 1);
+    }
+}
+
+static void ice_usb_uart_rx_to_cdc(struct uart_wrap* uart) {
+    while(1) {
+        // this can be executed 2 times if the data is wrapped
+        unsigned int bufsize = rb_data_left_continuous(&uart->rx_buf);
+
+        if (bufsize == 0) {
+            // no data in buffer
+            break;
+        }
+
+        // returns how much data was actually written
+        bufsize = tud_cdc_n_write(uart->itf, rb_get_read_ptr(&uart->rx_buf), bufsize);
+
+        if (bufsize == 0) {
+            // cdc buffer full
+            break;
+        }
+
+        // ack data that was sent to cdc
+        rb_read_ack(&uart->rx_buf, bufsize);
+    }
+
+    tud_cdc_n_write_flush(uart->itf);
+}
+
+
+// ------------------------------- Tx functions -------------------------------
+// cdc -> rb
+//  rb -> uart
+
+static void ice_usb_uart_tx_from_cdc(struct uart_wrap* uart) {
+    while(1) {
+        unsigned int bufsize = rb_space_left_continuous(&uart->tx_buf);
+        if (bufsize == 0) {
+            // no buffer space left
+            break;
+        }
+
+        bufsize = tud_cdc_n_read(uart->itf, rb_get_write_ptr(&uart->tx_buf), bufsize);
+        if (bufsize == 0) {
+            // cdc buffer empty
+            break;
+        }
+
+        rb_write_ack(&uart->tx_buf, bufsize);
+    }
+}
+
+static void ice_usb_uart_tx(struct uart_wrap* uart) {
+    // this could be done by DMA
```

Comment:
what about `uart_write_blocking()`?
Now you have a nice ring buffer API that can get all contiguous data buffers.
It does not use DMA either though:
https://github.com/raspberrypi/pico-sdk/blob/efe2103f9b28458a1615ff096054479743ade236/src/rp2_common/hardware_uart/include/hardware/uart.h#L459